### ISSUE TITLE
Watch external file changes

### ISSUE THREAD

**Issue by x128** [arvidn/libtorrent#262]
If a user edits/removes a partially downloaded file, libtorrent won't tell about the missing pieces. It will look like it's okay.

In my application, I can't monitor these external changes via WinAPI's FindFirstChangeNotification and Cocoa NSEvents, since they would also grab the changes made by libtorrent: how can I distinguish these changes?

**Comment by arvidn** [comment]
that's a good question. Is there a way to distinguish them at all? I can imagine using a heuristic of a "dead period" after issuing a write, but that wouldn't be fully accurate. Assuming there's no more information of which process issued a write or modification, it's not obvious whether it's possible.

**Comment by arvidn** [comment]
one think libtorrent does to protect against this is to open file in exclusive mode (this is optional, called "lock_files" or something like that, disabled by default)

**Comment by Falcosc** [comment]
@x128 did you find a workarround or do you have an idea to improve libtorrent to solve this issue?

**Comment by x128** [comment]
@Falcosc I have no idea except implementing "not a fully accurate" solution.

At least it would be accurate for the "file removed" case.

**Comment by Falcosc** [comment]
Does the exclusive open file mode solve your problem? And if you don't lock the files, missing files are already handled if somebody request a piece.

The only way to check the files is to trigger a hash check regulary, this could be done over api.

Or did you expect something else with this ticket?

**Comment by x128** [comment]
@Falcosc I don't expect anything, because I don't work on P2P software anymore. :)

I just believe that:
1. Exclusive mode is not good, because user (or maybe some antivirus app, or some script) should have a right to remove one of the downloaded files (accidentally or not).
2. Libtorrent should try to notice that and show some error state, and/or launch the hash check.
3. Anyway, it won't be accurate unless you dig it to the OS kernel level.

**Comment by Falcosc** [comment]
I guess we don't want to implement a listener for filesystem events because this woul be to complex for a plattform independend solution.

But it would be possible to check the file properties with periodically filesystem requests. This kind of feature would be a little bit expensive and raise the complexity.
I would like to tell the libtorrent users to check for file events by his own and manually trigger a hash check on file change/remove because most of them didn't support as many plattforms as libtorrent and it would be alot more easy for them to do this.
We already have anything exposed at the iterface make the libtorrent user able to implement it, or did we need to enhance something at the interface?
I would just extend the documentation for interface users to tell them they need to implement a plattform dependend solution for triggering a check based on changes/remove files.

**Comment by x128** [comment]
> I guess we don't want to implement a listener for filesystem events because this would be to complex for a plattform independend solution.

Your code is complex as long as you violate SOLID principles. I've implemented a few cross-platform solutions relying on **very** different OS entities. I'm sure that Arvid also has such solutions in libtorrent.

Libtorrent users cannot implement even a "not very accurate" solution mentioned above, because they have no idea about libtorrent's filesystem activity. Exposing it from the lowest layer of abstraction (like fwrites) to the highest one ("libtorrent user" interface) seems kind of crazy for me.

**Comment by Falcosc** [comment]
I forgot about changes during the download state you mentioned in your first post, you are right. That could not be exposed. I did just thought about file changes/remove in finish state.

But I guess there could be easy solution without implemnting all plattform dependent filesystem events.
We could make sure that it does not crash if we loose pieces on disk and if we enter finish state we would do the full hash check to make sure that we have everything.

I would like to have the following fallback as well for the finish state during uploading:
- if somebody request a piece which is missing/wrong on disk we remove set this piece to unfinished
- if we don't have a hashcheck during upload we could make it optional

For changes in finish state which did not get touched by piece requests, somebody need to implement a plattform dependend filesystem event listening. This is less importend then the handling on stateswitch to finish and less importend then the check during uploading. If somebody care about not requested files he could trigger a hash check based on filesystem events after torrent finished date

Does this new suggestion cover everything or did I forgett anything else?

In summary we could add or test the following functionallity to implement this enhancement:
- full hash check on state switch to finsih
- piece hash check on unpload (make it optional if we don't check them currently to not decrese performce)
- expose become_finished time point (don't need to store it in fast resume because we do hash checks on session creation with resume files)
- document how we handle filesystem changes and tell there that we do only check them on finish and on piece upload (if user want to do hash checks on piece upload)

**Comment by arvidn** [comment]
at the very least, it seems reasonable to, when detecting that a file is missing, all pieces overlapping that file are marked as "don't-have". The question then is, would those pieces be attempted to be re-downloaded? If not, should the file automatically be set to priority 0 (i.e. don't download)?

@Falcosc if you'd be willing to write a few tests covering this feature, I could implement it in libtorrent

**Comment by Falcosc** [comment]
Nice idea for cowork on this.

I did habe the following improvement
- Check upload hash only it Piece is not in cache.

I'm not sure about the target Städte after find an missmatch. But we could trigger a filesystem size check and existence check if wie found a wrong Piece if the file is gone then wie should stop the torrent.

**Comment by arvidn** [comment]
there is a feature called "seed mode", which is meant to avoid the initial hash check of a torrent when seeding. instead pieces are hash checked lazily, the first time it's requested. I don't think there's any need to build something that mimics or overlaps with this.

I think the most bang for the bucks is to simply detect missing files (or truncated files when reading past the end) and handle that gracefully, rather than just stopping the torrent with an error (which I believe is done now).

I think errors on writes should still be considered fatal and stop the torrent.

**Comment by Falcosc** [comment]
In downloading state:
- listen only on exceptions? **Not sure if we should avoid additional checks if we don't get IO problems**
-- write error: use the existing handling with upload only mode
-- errors which could be treated as missing file (if this kind of errors exist): stop torrent
- writing a changed file where we didn't get io errors: just go on and make sure that we have a full hash check on stateswitch to finish to take care of missing or wrong pieces before we declare it as finished
- detect missing files without IO error (would be a additional check):
-- I guess we can't just check the filesystem in downloadmode because there could be wired file states if user does something with the file during a write action
-- stop torrent if all is missing like on suggested handling of IO errors with missing files

In uploading state:
- listen on exceptions, no forced checks or only optional forced checks
-- read exception: mark piece as missing and trigger a filesystem parameter check
-- if file is empty or missing: stop the torrent
-- if file is not empty: **not sure what is a nice handling for this case because maybe there could be a temporarly read error?**
- read wrong blocks without get IO errors (for example the file did got overwritten by a file with same name): I guess a permanent hash check on ever not-cached piece request could help here. Optional of course for high performance seeders but enabled by default because normal users have less upload then CPU power. On hash check error: trigger advanced file check with size and maybe something else. Stop if the advanced file check declare it as not same file otherwise enter download mode which triggers a full hash check if we reach 100% again

**Comment by arvidn** [comment]
I don't really understand what point you're making. my opinion is still what I stated in https://github.com/arvidn/libtorrent/issues/262#issuecomment-272305832

I take it when you say "exception", you don't mean C++ exceptions, but errors returned by file I/O operations (which you don't really listen for, so much as handle).

with all those different cases, how many tests do you figure you would need to cover them all? I suspect an overwhelming number, not to speak of how complex the documentation would be (because it's not obvious that the behavior you're describing is intuitive).

**Comment by Falcosc** [comment]
I just want to make detail suggestions for the short comment "simply detect missing files and handle that gracefully"

The suggestions is idea to not decrease libtorrent performence. Instead of creating new checks on good usecases we wait for errors and if we got an error we spend some cpu time to check what is going on.

Test would be:

1. rename during download: expect ?
2. delete during download: expect ?
3. update file during download (I will update a text file of a multifile torrent because this is the most common usecase of fileschanges on accident): expect piece drop on last tick and repeat the piece download to get finished torrent afterwards
4. clear file during download: expect like 3 to make it more easy and less komplex
5. rename file after finish during seeding: expect ?
6. delete file after finish during seeding: expect ?
7. update file after finish during seeding: expect piece drop on request with rentering download mode
8. clear file after finish during seeding: expect like 7 to make it more easy and less komplex
9. file changes after finish during idle (all 4 changes in one test): expect no handling because checks during idle are wasteful with comment that this expection could be changed if filesystem events are implemented

Do you know how the used io classes handle a moved file? Maybe it's os specific. I just remember usecases from other software where a file move result into a new empty file at source. And I don't know if this is the default way of some io libs or if it the error handling of the developers. I just want to say we should use the default behaivor of the io functions to do as less addiotional checks as possible for the good usecases.
For that reason I create rename and delete tests because it is maybe handled not in same way.

I would like to add expectations to the test suggestions. But my expectations depend on the solution with the most less performance decrease. For example 1: should I exepect a stoped torrent because io classes give us an error which we could handle with a stop. Or should I exepect a progress drop from 99% to 49% because we did not get an error and download untill we check the whole file just before reaching finishstate

solution of 7 could be a little bit expensive. Maybe an optional check hash always during upload which checks hash evertime a seed piece got loaded into cache. Or lazy check like seed_mode with extension to reset check progress on file modification date change (file modification date check is done evertime a seed piece got loaded into cache)

**Comment by arvidn** [comment]
I think (5) should be treated as (6) and the only cases to change the current behavior for.

The existing behavior is to stop the torrent. I think a reasonable new behavior would be to change state from seeding to finished and set the file priority to 0 for the file that disappeared.

there's no practical way of detecting an external process corrupting the file, short of always checking hashes while seeding (which isn't really practical).

The main risk is to not break the case where people download to an external drive that happens to be unmounted. Once it comes back, it would be good to do something sensible

**Comment by Falcosc** [comment]
I guess we get an IO error on 
- reading missing filesystem
- reading reading permission failed if an folder of open file get altered permissions
- reading moved/deleted file
Then we would have everthing we need which means no perfomance decrease for the errorless usecases.
I would expect finished and set the file priority to 0 for IO errors on read (5) and (6)

I know there's no practical way of detecting an external process corrupting the file to handle (7)
Always checking is the only way but the amount of checks could be reduced
- do it only for non cached pieces
- use seed mode lazy check but extend it with a check progress reset on modification change (use memory for another timepoint and check modification date on non cached piece requests)

How does seed mode handle wrong hashes? Do we need to change the expectation of (7) or the implementation

**Comment by arvidn** [comment]
one risk that makes me a bit hesitant is, if the whole download directory is unmounted, you would end up having all your torrents have all their files disappear and have priority 0. and it would not be obvious how to restore your state once you re-mount the drive.

In fact, I think this would be a greater loss than the gain of detecting files having been removed.

if see mode gets a hash check failure, it leaves seed mode and triggers a full recheck

**Comment by Falcosc** [comment]
Unmounted files disappear on the first seed request. We could make it able to disable if somebody mind it would be bad but enable it by default. I'm fine with it because if the user could force recheck to restore the state. And users could just restart the session, then I guess everything gots an recheck and should be reset or doesn't it with the suggested prio 0?

And for the last point, what about if I add a small detail
if seed mode gets a hash check failure (no IO errors), it leaves seed mode, download the missing piece and then it triggers a full check on finish action.

**Comment by arvidn** [comment]
the priority would not automatically be restored when a file re-appeared. And re-checking files is actually quite expensive. I don't think that's a reasonable expectation. Furthermore, I would expect the "unmounted download directory" to be a far more common use case than the "moving files manually while seeding them"-use case.

The original intention of seed mode is that encountering a hash check should be very rare, so the exact behavior when it happens is not all that important (as long as it's correct obviously). It's simpler to leave seed mode immediately, recheck and then download the missing pieces. Otherwise there need to be a seed-mode state where we're not a seed, which is complex.

when downloading (or specifically when writing to files), a missing file is not an error. The file will simply be created.

As for the original intent of this ticket, I don't think it's possible to reliably implement this. It's possible to detect updates to the files, them being moved etc. But by then the disk threads may already experienced errors. To know whether those errors were caused by the external file operation or not, and potentially ignore them and update our path to where the file is. All "atomically" with the disk thread, is very complicated and error prone.

**Comment by Falcosc** [comment]
Then I will expect for 1 and 2 that we just go on with downloading and get a progress drop during finish check.

> The original intention of seed mode is that encountering a hash check should be very rare, so the exact behavior when it happens is not all that important (as long as it's correct obviously). It's simpler to leave seed mode immediately, recheck and then download the missing pieces. Otherwise there need to be a seed-mode state where we're not a seed, which is complex.
I think I get it, we don't have a seed request based hashcheck for download mode? That would have a very bad impact for the usecase 1 and 2 because if we don't know that we are missing 50% of the file because it got deleted mid download.
Do you have an idea to solve this? I guess it's only possible with permanent hash checks of uncached piece requets in download state. something like pessimistic download mode which is expensive but failsave for unexpected user actions. And the finish state hash check take care of pieces with did not get requested during pessimistic download mode.

**Comment by arvidn** [comment]
the one way I can think of solving that would be:

if we discover that a file we're writing to doesn't exist, before creating it, make sure we don't have any other pieces in that file.

The problem, though, is we may have pieces in the part file. It would also make the interaction between the torrent object and the disk thread more complex. It's not obvious to me that supporting any of these cases are worth the added complexity.

I don't see a compelling use case.

**Comment by Falcosc** [comment]
Usecase is interacting with not finished or partial finished downloads. That could be quite common.

Over all I would even take the risk of seed broken pieces during download for this rare usecase.
What did you think about a pessimistic download mode? I don't like aktive reacting on filesystem changes, too because there is no useable solution for it. But hashcheck on all requests during download mode should be much more easy to implement. Don't do advanced reactions on failed hashes in pessimistic download mode, just remove them. The result would be a full downloaded file at the end. Even if the user overwrites it with the old content, the finish check would discover missing pieces.

**Comment by arvidn** [comment]
when you add a torrent and some files exist already there are two cases that are handled:

1. if you have a fast-resume data file, there's some sanity checking of it and then accepted as the current state.
2. a full check of the files is performed to build the state

I would expect those cases to cover the majority of "interacting with not finished or partial finished downloads".

Maybe if seed_mode could be generalized it would make sense to be able to turn it on for downloads too. Then it would probably make sense to add a probabilistic factor too. You may not want to (redundantly) check the hash of every piece all the time, but with some likelihood. You also probably would want some kind of cool-down, to not check the same piece several times per second just because you're getting lots of requests to it.

Fundamentally though, there are so many other things that users actually do care about that could be done instead of this. Also, I wouldn't want to jam more complexity into the torrent object. For this to be feasible, I think seed mode needs to be factored out into a separate class and be unit testable independently.

**Comment by Falcosc** [comment]
I guess seed mode logic does not work in downloading state. Because once you did check the hash it will be never checked again.
And you can't reset the check state based on modification dates because we modified the file during download.
Or do you meen to add a time based check state reset for seed mode?
- before finish: time based lazy hashcheck progress reset before finish
- after finish: file modification date change based lazy hashcheck progress reset

If something is to difficoult: at the end we may not need to care about broken pieces during download. Would be ok for me to delivery broken pieces and wait until we have mark everthing as have to trigger the finish full check.
